Frontend video image validation

mushtaqak opened the pull request:
This PR adds frontend validation to video images on video uploads page.

[Sandbox](https://studio-videoimage.sandbox.edx.org/videos/course-v1:edx+VideoTest+2017_R1)

muhammad-ammar commented on cms/static/js/views/video_thumbnail.js:
```diff
@@ -162,9 +205,15 @@ define(
-                data.submit();
+                var errorMessage = this.validateImageFile(data.files[0]);
+                if (!errorMessage) {
+                    data['global'] = false;   // Do not trigger global AJAX error handler
```
you can also use `data.global = false`

muhammad-ammar commented on cms/static/js/views/video_thumbnail.js:
```diff
@@ -216,6 +278,36 @@ define(
 
+            validateImageFile: function(imageFile) {
+                var errorMessage = '';
+                var self = this,
```
A single `var` is enough.

muhammad-ammar commented on cms/static/js/views/video_thumbnail.js:
```diff
@@ -216,6 +278,36 @@ define(
+                    fileType;
+
+                fileName = imageFile.name;
+                fileType = fileName.substr(fileName.lastIndexOf('.'));
```
I think we should compare content type instead of file extension.

muhammad-ammar commented on cms/static/js/views/video_thumbnail.js:
```diff
@@ -216,6 +278,36 @@ define(
+                fileName = imageFile.name;
+                fileType = fileName.substr(fileName.lastIndexOf('.'));
+
+                if (!_.contains(self.getVideoImageSupportedFileFormats().machine, fileType)) {
```
Same as above.

muhammad-ammar commented on cms/static/js/views/video_thumbnail.js:
```diff
@@ -216,6 +278,36 @@ define(
+                if (!_.contains(self.getVideoImageSupportedFileFormats().machine, fileType)) {
+                    errorMessage = gettext(
+                        'The selected image contains unsupported image file type. ' +
+                        'Supported file formats are {supportedFileFormats}.'
```
This message is different from backend validation message.

srpearce replied:
This image file type is not supported. Supported file types are .gif, .jpg, and .png.

muhammad-ammar commented on cms/templates/js/video-thumbnail.underscore:
```diff
@@ -1,15 +1,17 @@
-        <input id="thumb-<%- videoId %>" class="upload-image-input" type="file" name="file" accept=".bmp, .jpg, .jpeg, .png, .gif"/>
+        <input id="thumb-<%- videoId %>" class="upload-image-input" type="file" name="file"/>
         <label for="thumb-<%- videoId %>" class="thumbnail-action">
-            <span class="action-icon" aria-hidden="true"><%- actionInfo.icon %></span>
```
Why we deleted this? We need icon for `Edit Thumbnail`.

mushtaqak commented on cms/static/js/views/video_thumbnail.js:
```diff
@@ -174,10 +223,23 @@ define(
+
+            showErrorMessage: function(errorText) {
+                this.readMessages([gettext('Video image upload failed'), errorText]);
+                HtmlUtils.setHtml(
```
@muhammad-ammar There is a bug here, after testing many times found that when an error has been shown, we can't upload image again. After debugging found out that we are updating HTML, so view has changed so event is not caught by file upload handler. I think,  this would mean that I won't be able to tag doc team on sandbox.

mushtaqak replied:
this is fixed now :)

mushtaqak commented:
@srpearce Can you please look at image validation messages ? and user facing strings ?

srpearce commented on cms/static/js/spec/views/video_thumbnail_spec.js:
```diff
@@ -149,38 +195,152 @@ define(
+                    failureData = {
+                        jqXHR: {
+                            responseText: JSON.stringify({
+                                error: 'The selected image contains unsupported image file type.'
```
This image file type is not supported. Supported file types are .gif, .jpg, and .png.

srpearce commented on cms/static/js/spec/views/video_thumbnail_spec.js:
```diff
@@ -149,38 +195,152 @@ define(
+                expect(videoThumbnailView.readMessages).toHaveBeenCalledWith(['Video image upload completed']);
+                videoThumbnailView.imageUploadFailed({}, failureData);
+                expect(videoThumbnailView.readMessages).toHaveBeenCalledWith(
+                    ['Video image upload failed', 'The selected image contains unsupported image file type.']
```
Could not upload the video image file.

This image file type is not supported. Supported file types are .gif, .jpg, and .png.

srpearce commented on cms/static/js/spec/views/video_thumbnail_spec.js:
```diff
@@ -149,38 +195,152 @@ define(
+                    .fileupload('add', {files: [createFakeImageFile(VIDEO_IMAGE_MIN_BYTES, 'mov/mp4')]});
+
+
+                errorMessage = 'The selected image contains unsupported image file type. Supported file formats are ' +
```
This image file type is not supported. Supported file types are .gif, .jpg, and .png.

srpearce commented on cms/templates/js/video-thumbnail.underscore:
```diff
@@ -9,7 +9,10 @@
         <span class="requirements-text-sr sr">
-            <%- gettext('Recommended image resolution is 1280x720 pixels and format must be one of .jpg, .png or .gif') %>
+            <%- edx.StringUtils.interpolate(
+                gettext("Recommended image resolution is 1280x720 pixels, maximum image file size should be {maxFileSize} and format must be one of {supportedImageFormats}."),
```
Can you give me a screen shot of where this message appears?

mushtaqak replied:
This is only for screen reader.

<img width="683" alt="screen shot 2017-06-01 at 4 27 30 pm" src="https://cloud.githubusercontent.com/assets/6991154/26677916/6a6e60a0-46e7-11e7-848d-5529b9a36b81.png">

mushtaqak commented:
@muhammad-ammar @srpearce address feedback. Sandbox updated.

mushtaqak commented:
@cptvitamin @roderickmorales Can you please review video image feature. Sandbox at https://studio-videoimage.sandbox.edx.org/videos/course-v1:edx+VideoTest+2017_R1

muhammad-ammar commented on cms/static/js/spec/views/previous_video_upload_spec.js:
```diff
@@ -14,7 +14,21 @@ define(
+                                '.bmp2': 'image/x-ms-bmp',   // PIL gives x-ms-bmp format
+                                '.jpg': 'image/jpeg',
+                                '.jpeg': 'image/jpeg'
+                            }
```
Why we need to set all these values to videoImageSettings? Can't we set it to empty? These settings are only needed and tested in `video_thumbnail_spec.js` so why we need these here?

muhammad-ammar commented on cms/static/js/spec/views/previous_video_upload_list_spec.js:
```diff
@@ -25,7 +25,21 @@ define(
+                                '.bmp': 'image/bmp',
+                                '.bmp2': 'image/x-ms-bmp',   // PIL gives x-ms-bmp format
+                                '.jpg': 'image/jpeg',
+                                '.jpeg': 'image/jpeg'
```
Why we need to set all these values to videoImageSettings? Can't we set it to empty? These settings are only needed and tested in `video_thumbnail_spec.js` so why we need these here?

srpearce commented:
@mushtaqak @roderickmorales - I'm not sure if this is the best place to leave this comment, but I've noticed that when I resize the page, the Video ID column shrinks, but the Name column doesn't. Since the video name is frequently shorter than the video ID, it might make sense to change it so that the Video ID column doesn't shrink so much.

![screenshot 25](https://cloud.githubusercontent.com/assets/3493253/26730234/375fcd92-477e-11e7-92bc-eeaf43fdb1cd.png)

mushtaqak commented on cms/static/js/spec/views/video_thumbnail_spec.js:
```diff
@@ -149,38 +186,152 @@ define(
+                $el.find('.upload-image-input')
+                    .fileupload('add', {files: [createFakeImageFile(VIDEO_IMAGE_MIN_BYTES - 10)]});
+
+                errorMessage = 'The selected image must be larger than ' +
```
@muhammad-ammar I am verifying meesage from the view below, this is just a a variable to hold expected error message

mushtaqak commented on cms/static/js/spec/views/video_thumbnail_spec.js:
```diff
@@ -149,38 +186,152 @@ define(
+                    .fileupload('add', {files: [createFakeImageFile(VIDEO_IMAGE_MAX_BYTES + 10)]});
+
+                errorMessage = 'The selected image must be smaller than ' +
+                    videoThumbnailView.getVideoImageMaxSize().humanize + '.';
```
@muhammad-ammar I am verifying meesage from the view below, this is just a a variable to hold expected error message

mushtaqak commented on cms/static/js/spec/views/video_thumbnail_spec.js:
```diff
@@ -149,38 +186,152 @@ define(
+                    .fileupload('add', {files: [createFakeImageFile(VIDEO_IMAGE_MIN_BYTES, 'mov/mp4')]});
+
+
+                errorMessage = 'This image file type is not supported. Supported file types are ' +
```
@muhammad-ammar I am verifying meesage from the view below, this is just a a variable to hold expected error message

mushtaqak commented on cms/static/js/views/video_thumbnail.js:
```diff
@@ -64,13 +70,49 @@ define(
+                return {
+                    humanize: supportedFormats.slice(0, -1).join(', ') + ' or ' + supportedFormats.slice(-1),
+                    machine: _.values(this.videoImageSettings.supported_file_formats).sort()
+                };
```
Even though we send sorted format from backend, `reject(_.keys(this.videoImageSettings.supported_file_formats)` changes the array and returns unsorted array. Therefore we need to do this here.

mushtaqak commented on cms/static/js/views/video_thumbnail.js:
```diff
@@ -191,7 +239,11 @@ define(
+                // When we had error, focused effect was not wearing off after hover out.
+                if (!$(this.$el.parent()).hasClass('has-thumbnail-error')) {
+                    this.$('.thumbnail-wrapper').addClass('focused');
+                }
```
we need this because When we had error, focused effect was not wearing off even after hover out on the image containing error. `this.$('.thumbnail-wrapper').addClass('focused'); ` adds to all elements, however, here, we want to skip those with error so as to avoid focussed effect issue.

mushtaqak commented on cms/static/js/views/video_thumbnail.js:
```diff
@@ -204,10 +256,17 @@ define(
+                    HtmlUtils.setHtml(
+                        this.$('.thumbnail-action .action-icon'),
+                        HtmlUtils.HTML('')
+                    );
```
We are adding icon except when action is error. This is in regard to final UI mockups.

mushtaqak commented:
@Qubad786 @muhammad-ammar addressed feedbacks

mushtaqak commented:
Sandbox updated : https://studio-videoimage.sandbox.edx.org/videos/course-v1:edX+DemoX+Demo_Course

mushtaqak commented on cms/static/js/views/video_thumbnail.js:
```diff
@@ -162,9 +204,16 @@ define(
-                this.readMessages([gettext('Video image upload started')]);
-                this.showUploadInProgressMessage();
-                data.submit();
+                var errorMessage = this.validateImageFile(data.files[0]);
```
A miner bug just found: Need to remove error message after we select new image.
@muhammad-ammar @muzaffaryousaf FYI

roderickmorales commented:
@mushtaqak here are the typography changes that I mentioned to you earlier
![thumnail-hover-text](https://cloud.githubusercontent.com/assets/1260455/26794981/d86dbe72-49f1-11e7-86f1-c5a667e29a2e.png)

cptvitamin commented:
👍

muhammad-ammar commented on cms/static/js/views/video_thumbnail.js:
```diff
@@ -216,6 +278,145 @@ define(
+                } else if (imageFile.size < this.getVideoImageMinSize().machine) {
+                    errorMessage = StringUtils.interpolate(
+                        // Translators: minFileSizeInKB will be like 2 KB.
+                        gettext('The selected image must be smaller than {minFileSizeInKB}.'),
```
@mushtaqak `The selected image must be smaller than` -> `The selected image must be larger than`

Credit goes to @Qubad786 for catching this :)

mushtaqak replied:
Haha I had fixed this earlier :P

mushtaqak commented:
@roderickmorales Addressed feedback, please check https://studio-videoimage.sandbox.edx.org/videos/course-v1:edX+DemoX+Demo_Course

@muhammad-ammar @Qubad786 Final reviews?

mushtaqak commented:
@sstack22 @roderickmorales @srpearce Final reviews for video image ?

Updated sandbox : https://studio-videoimage.sandbox.edx.org/videos/course-v1:edX+DemoX+Demo_Course

@muzaffaryousaf FYI

srpearce commented:
@mushtaqak I'm still seeing the crowding issue on both Chrome and Firefox, and other doc team members are seeing it as well:

Firefox
![videouploads_firefox](https://cloud.githubusercontent.com/assets/3493253/26835146/92d116ac-4aa5-11e7-9817-4c20ba78562b.png)

Chrome
![videouploads_chrome](https://cloud.githubusercontent.com/assets/3493253/26835321/1175d9b6-4aa6-11e7-83a5-a34aff156166.png)

srpearce commented:
From a UI text standpoint, everything is good to go. 

It would also be nice if the Name column resized/wrapped instead of the Video ID column (see image below), but I'll let @roderickmorales speak to that.

![screenshot 40](https://cloud.githubusercontent.com/assets/3493253/26835594/a10da734-4aa6-11e7-9efe-deaeb5535f4e.png)

mushtaqak commented:
@srpearce I think this page is not made responsive that is why it happens like that. We can make that better but that would be out of the scope of this PR, I suppose. @muzaffaryousaf what do you think ?

sstack22 commented:
👍  looks good to me, @mushtaqak , @muzaffaryousaf  - from roll out, have we discussed with DevOps? It would also be good if this functionality landed at the same time the default image did.

muzaffaryousaf commented:
@sstack22 We're working with DevOps to setup s3 integration for this feature for stage, edge and production environments. Also, we're planning to do a bug bash for this feature tomorrow. Is that what you were asking ? 

> It would also be good if this functionality landed at the same time the default image did.

The functionality is being done from our side but we're waiting for s3 integration to be completed so that we can do end-2-end testing.

muzaffaryousaf commented:
@mushtaqak @srpearce That's not part of this PR/work. Maybe we can address this issue when we'll start work on video uploads re-design project.

muhammad-ammar commented:
👍 after build is green.

Qubad786 commented:
jenkins run js

mushtaqak commented:
jenkins run bokchoy

muzaffaryousaf commented:
> @mushtaqak here are the typography changes that I mentioned to you earlier
![thumnail-hover-text](https://cloud.githubusercontent.com/assets/1260455/26794981/d86dbe72-49f1-11e7-86f1-c5a667e29a2e.png)

@mushtaqak I'm seeing that you've not addressed @roderickmorales 's feedback or maybe the sandbox which you attached is not up-to-date. Can you make these changes ? Also, if you've updated an existing sandbox then probably you've to compile the assets first.
